### ISSUE TITLE
documentation for costing changes in plutus v3

### ISSUE THREAD

**Issue by michele-nuzzi** [IntersectMBO/plutus#6319]
### Describe the feature you'd like

recently I was testing some plutus v3 contracts.

at submission I noticed that lates costs in sancho are an array of length 251 instead of 233 expected

it does not appears that new builtins where added, instead it seems that some have changed their formula

I have notice `quotientInteger` and `remainderInteger` now quadratic in both argument

I also noticed that 2 parameters are negative, can you confirm is not a bug?

it would be nice to have clear how to use the new costs introduced and which builtins have been changed

**Comment by michele-nuzzi** [comment]
additionally, is there anywhere I can join to get notified about these changes as they happen (some discord server or similar)?

**Comment by michele-nuzzi** [comment]
if the negative integer is not a bug I assume it is encoded as negative integer on the respective language view for the script data hash calculation in the various transactions

**Comment by effectfully** [comment]
@kwxm please response to these once you have some time.

**Comment by michele-nuzzi** [comment]
Thank you very much

If an explicit mapping from the position of the cost in the array to the cost property name is also available that would be of great help

**Comment by kwxm** [comment]
@michele-nuzzi Apologies for the delay in getting back to you.

> recently I was testing some plutus v3 contracts.

> at submission I noticed that lates costs in sancho are an array of length 251 instead of 233 expected

> it does not appears that new builtins where added, instead it seems that some have changed their formula

> I have notice quotientInteger and remainderInteger now quadratic in both argument

> I also noticed that 2 parameters are negative, can you confirm is not a bug?

That's correct.  We've updated the cost model for PlutusV3 to account for improvements to tthe evaluator and the addition of some new built-in functions and we've also improved the precision of costing for some functions, including the integer division ones.  These changes should reduce the CPU costs of a typical script by 20-30%; memory costs will be essentially unchanged.  The negative numbers are correct,  and we enforce a minimum value to make sure that the cost of calling the function is never negative.  You can  see the code that does the calculation [here](https://github.com/IntersectMBO/plutus/blob/ffe41b864ea018e188ab9e0295cf6ad7e180844a/plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostingFun/Core.hs#L378).

We're aware that there's not much public documentation about the  cost model, so I've attached a draft of an explanation:  this may still contain some typos, so read it with caution.  I hope it'll still be of some help though.

[cost-model-overview.pdf](https://github.com/user-attachments/files/16422671/cost-model-overview.pdf)

**Comment by kwxm** [comment]
> additionally, is there anywhere I can join to get notified about these changes as they happen (some discord server or similar)?

Yes.  [Intersect](https://www.intersectmbo.org/) has a Discord server where you could talk to us a bit more directly, in the `plutus-working-group` channel.  I think that you may have to be a member of Intersect to be able to access that though, and there's a small fee to join ($10 I think).

**Comment by kwxm** [comment]
> If an explicit mapping from the position of the cost in the array to the cost property name is also available that would be of great help.

That's determined by the ordering of the parameters [here](https://github.com/IntersectMBO/plutus/blob/ffe41b864ea018e188ab9e0295cf6ad7e180844a/plutus-ledger-api/src/PlutusLedgerApi/V3/ParamName.hs#L17).

**Comment by michele-nuzzi** [comment]
Thank you @kwxm, the ordering helped a lot, if this issue is not useful to you guys I think we can close it.

**Comment by effectfully** [comment]
Great then! Glad we were able to help. Thank you @kwxm!

**Comment by nielstron** [comment]
Hi, sorry to ping again in this thread. The cost-model-overview is super helpful. Figure 1 references a cost model file that seems similar to previously posted cost model files for mainnet in the Cardano Book (https://book.play.dev.cardano.org/cost-model.html) however the most recent version (2024) contains only a list of integers and no PlutusV3 instead of the descriptive key-value pairs. Is there anywhere I could obtain the key-value version or alternatively a mapping from the list of integers to the key-value version (machine readable!)

**Comment by kwxm** [comment]
> I am interested in a way to obtain the "flattened cost model entries" of the current network parameters.

That's actually a little difficult.  I should be able to supply you with this, but it'll take a little time. I'll get back to you on that later.

**Comment by michele-nuzzi** [comment]
@nielstron I got the flat using this script

https://github.com/HarmonicLabs/cardano-costmodels-ts/blob/main/scripts%2FprintFlatV3.ts

**Comment by nielstron** [comment]
Thanks @michele-nuzzi , thats a great source. I implemented something similar myself before but the additional pointer to the enum/order of params is very helpful! This way I can also extract the flat form easily from the propagated parameter list on chain

**Comment by kwxm** [comment]
Thank you for that @michele-nuzzi: that's saved me some trouble.

**Comment by kwxm** [comment]
The cost model document is actually incorrect.  It says that the cost model parameters are stored on the node in the tagged form (`"equalsInteger-cpu-arguments-intercept": 51775` etc).  I think that that used to be the case but now only a list of untagged integers is stored.  These have to be in a specific order (defined in [this file](https://github.com/IntersectMBO/plutus/blob/master/plutus-ledger-api/src/PlutusLedgerApi/V3/ParamName.hs) and similar ones) and when the evaluator receives the list of integers from the ledger it has to re-tag them in order to update the evaluator's internal cost model correctly. We'd like to simplify this process but it's diffiult to do that at the moment for reasons of backward compatibility.

Note also that the current on-chain cost model parameters contain entries for new built-in PLC functions like `xorByteString` and `ripemd_160` (the "Batch 5" functions in the Plutus Core specification).  These functions can't be used on the chain yet: they'll only be [enabled in PV10](https://github.com/IntersectMBO/plutus/blob/5460840672e9cb86b4c61d12903fde42c2de9ef8/plutus-ledger-api/src/PlutusLedgerApi/Common/Versions.hs#L123), which should happen at the next hard fork.
